### ISSUE TITLE
Several 100 Threads hanging on epollwait

### ISSUE THREAD

**Issue by devilchn** [eclipse/jetty.project#2326]
similar to https://github.com/eclipse/jetty.project/issues/1033

jetty: 9.4.5.v20170502
jvm: 8.1.033 25.51-b02
OS: linux, aix

I have a lot of threads hanging on epollwait (AIX hangs on poll0):

   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.EPollArrayWrapper.epollWait(JIJI)I(Native Method)
	at sun.nio.ch.EPollArrayWrapper.poll(J)I(EPollArrayWrapper.java:269)
	at sun.nio.ch.EPollSelectorImpl.doSelect(J)I(EPollSelectorImpl.java:93)
	at sun.nio.ch.SelectorImpl.lockAndDoSelect(J)I(SelectorImpl.java:86)
	- locked <0x00000000a45d9448> (a sun.nio.ch.Util$3)
	- locked <0x00000000a45d9438> (a java.util.Collections$UnmodifiableSet)
	- locked <0x00000000a45d93f0> (a sun.nio.ch.EPollSelectorImpl)
	at sun.nio.ch.SelectorImpl.select(J)I(SelectorImpl.java:97)
	at sun.nio.ch.SelectorImpl.select()I(SelectorImpl.java:101)
	at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.select()Z(ManagedSelector.java:234)
	at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.produce()Ljava/lang/Runnable;(ManagedSelector.java:187)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce()V(EatWhatYouKill.java:211)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce()V(EatWhatYouKill.java:126)
	at org.eclipse.jetty.io.ManagedSelector$$Lambda$28/269758229.run()V(Unknown Source)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(Ljava/lang/Runnable;)V(QueuedThreadPool.java:672)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run()V(QueuedThreadPool.java:590)
	at java.lang.Thread.run()V(Thread.java:808)

This doesn't cause the CPU consumption to spike significantly as in most similar things I googled for. However, other tasks fail due to a lack of threads or time-out: 

   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(ZJ)V(Native Method)
	- parking to wait for  <0x00000000e9dec4f0> (a java.util.concurrent.CountDownLatch$Sync) //noone is holding that
	at java.util.concurrent.locks.LockSupport.parkNanos(Ljava/lang/Object;J)V(LockSupport.java:215)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedNanos(IJ)Z(AbstractQueuedSynchronizer.java:1037)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(IJ)Z(AbstractQueuedSynchronizer.java:1328)
	at java.util.concurrent.CountDownLatch.await(JLjava/util/concurrent/TimeUnit;)Z(CountDownLatch.java:277)
	at org.eclipse.jetty.io.ManagedSelector$CloseEndPoints.await(J)Z(ManagedSelector.java:719)
	at org.eclipse.jetty.io.ManagedSelector.doStop()V(ManagedSelector.java:114)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.stop()V(AbstractLifeCycle.java:89)
	- locked <0x00000000a16a8c28> (a java.lang.Object)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.stop(Lorg/eclipse/jetty/util/component/LifeCycle;)V(ContainerLifeCycle.java:142)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStop()V(ContainerLifeCycle.java:160)
	at org.eclipse.jetty.io.SelectorManager.doStop()V(SelectorManager.java:257)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.stop()V(AbstractLifeCycle.java:89)
	- locked <0x00000000a149fa10> (a java.lang.Object)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.stop(Lorg/eclipse/jetty/util/component/LifeCycle;)V(ContainerLifeCycle.java:142)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStop()V(ContainerLifeCycle.java:160)
	at org.eclipse.jetty.client.AbstractHttpClientTransport.doStop()V(AbstractHttpClientTransport.java:87)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.stop()V(AbstractLifeCycle.java:89)
	- locked <0x00000000a0a3b248> (a java.lang.Object)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.stop(Lorg/eclipse/jetty/util/component/LifeCycle;)V(ContainerLifeCycle.java:142)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStop()V(ContainerLifeCycle.java:160)
	at org.eclipse.jetty.client.HttpClient.doStop()V(HttpClient.java:254)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.stop()V(AbstractLifeCycle.java:89)
	- locked <0x00000000a0a3ae00> (a java.lang.Object)
...

From a look at a heap dump I was able to gather, that Selectors in the epollwait threads do not have a select Key attached. What are these threads doing? I have not found a thread that has a parent. 

It seems to be happening on somewhere between 1 in 10 and 1 in 1000 hosts where the application is running. A restart of the application does not remedy the issue for long. It is somewhat likely that the hosts that show this behavior have some network issues due to high load. However, I have been able to connect to them via SSL and the connection seemed stable.

**Comment by sbordet** [comment]
Threads waiting in the `select()` operation are normal.
You can control their number by configuring the number of selectors in the `ServerConnector`.
If you are using Jetty as a standalone server, the property to use is `jetty.http.selectors`.
The default is typically half the number of cores.
https://www.eclipse.org/jetty/documentation/current/configuring-connectors.html

**Comment by devilchn** [comment]
@sbordet 
No, it is not a standalone Jetty. There are 600 Threads waiting on select(). I got some thread dumps where everything is working fine. There are less than 5 threads on select().
Also it seems very strange to me that they have no select key attached.

**Comment by sbordet** [comment]
There seems to be `HttpClient` instances as well. Make sure you have only one (or few) and not many, as each `HttpClient` also uses NIO and has selector threads.
Also, make sure you don't use other libraries that may use NIO (e.g. Netty).

**Comment by sbordet** [comment]
@devilchn news ?

**Comment by sbordet** [comment]
Closing, lack of activity. Feel free to reopen if you have more information.

**Comment by ginkosen** [comment]
**jetty**: 9.4.30.v20200611 **jdk**:11.0.7

same issues with jstack

![image](https://user-images.githubusercontent.com/2233819/169328781-f9226bdd-d2ec-4ff6-a56d-5e4bc2280fdd.png)

**Comment by joakime** [comment]
@ginkosen what OS? 

And upgrade your Jetty version, version 9.4.30 is subject to several security vulnerabilities.
See: https://www.eclipse.org/jetty/security_reports.php

**Comment by devilchn** [comment]
We were not able to figure it out and moved to Apache HTTP client prior to the issue. So I was not able to follow up and I am sorry I never said so. A JVM expert did some analysis and pointed away from the JVM and we had it on multiple OS.
He said there are multiple "ManagedSelector" instances alive that no one uses. He was sure that this would be solvable with a good understanding of the Jetty Framework (which we also lacked).

**Comment by ginkosen** [comment]
@joakime  

centos7.9  

4 physical  cpu with 20 thread

1T memory

we upgrade jetty still not work.


we use jar decency and start with java main function

**Comment by ginkosen** [comment]
**supplementary instruction** 
centos7.9 
8 core 
16G memory

Services start normally

**Comment by sbordet** [comment]
@ginkosen please open a new issue and report more information.
Very likely you are not using `HttpClient` correctly, or you have a misconfiguration.
In the new issue, report a server dump, like explained here: https://www.eclipse.org/jetty/documentation/jetty-10/operations-guide/index.html#og-troubleshooting-dump

**Comment by ginkosen** [comment]
@sbordet 

Hi.Sorry for slow reply.

We figure out the problem.

When we disabled intel cpu HT, the problem is sloved.

**When we use the official demo, the program runs normally, this is the problem of our program**.
